src/repositories/request-repository.ts

```diff
@@ -202,165 +249,88 @@ export class RequestRepository extends Repository<Request> {
-              message: REQUEST_MESSAGES.conflictResolutionRejection,
-            }
-          )
-          .orWhere(
-            'request.status = :processingStatus AND request.updatedAt < :processingDeadline',
-            {
-              processingStatus: RequestStatus.PROCESSING,
-              processingDeadline: processingDeadline,
-            }
-          )
-          .orWhere('request.status = :pendingStatus', { pendingStatus: RequestStatus.PENDING })
-
-        // using conditions in allRequestsToAnchorQuery, retrieves up to streamLimit unique streams with their earliest request createdAt value
-        const rawStreamsToAnchor = await allRequestsToAnchorQuery
-          .clone()
-          .select(['request.streamId as sid', 'MIN(request.createdAt) as min_created_at'])
-          .groupBy('sid')
-          .orderBy('min_created_at', 'ASC')
-          // if 0 will return unlimited
-          .limit(maxStreamLimit)
-          .getRawMany()
-
-        // convert raw results to Request entities
-        const streamsToAnchor = transactionalEntityManager.getRepository(Request).create(
-          rawStreamsToAnchor.map((request) => ({
-            streamId: request['sid'],
-            createdAt: request['min_created_at'],
-          }))
-        )
+    const { connection = this.connection } = options
+    const now = new Date()
+    const anchoringDeadline = new Date(now.getTime() - this.config.maxAnchoringDelayMS)
+
+    return connection.transaction(
+      async (trx) => {
+        const streamsToAnchor = await findStreamsToAnchor(trx, maxStreamLimit, now)
 
-        // Do not anchor if there are no streams to anhor
+        // Do not anchor if there are no streams to anchor
         if (streamsToAnchor.length === 0) {
           logger.debug(`Not updating any requests to READY because there are no streams to anchor`)
           return []
         }
 
         // Anchor if we have enough streams or the earliest stream request is expired
-        const earliestIsExpired = streamsToAnchor[0].createdAt < anchoringDeadline
-        if (streamsToAnchor.length >= minStreamLimit || earliestIsExpired) {
+        const enoughStreams = streamsToAnchor.length >= minStreamLimit
+        const earliestIsExpired = streamsToAnchor[0].minCreatedAt < anchoringDeadline
+
+        if (enoughStreams || earliestIsExpired) {
           const streamIds = streamsToAnchor.map(({ streamId }) => streamId)
 
-          // using conditions in allRequestsToAnchorQuery, retrieves all requests associated with the retrieved streams
-          const requests = await allRequestsToAnchorQuery
-            .orderBy('request.createdAt', 'ASC')
-            .andWhere('request.streamId IN (:...streamIds)', { streamIds })
-            .getMany()
+          const requests = await findRequestsToAnchorForStreams(trx, streamIds, now)
 
-          const results = await this.updateRequests(
+          const updatedCount = await this.updateRequests(
```

Comment:
nitpick: consider a markRequestsReady helper that can handle the query plus checking if the updatedCount doesn't match the requests length